aten/src/ATen/core/ivalue_inl.h

```diff
@@ -680,42 +682,27 @@ inline IValue::IValue(c10::intrusive_ptr<ivalue::ConstantString> v)
 inline IValue::IValue(std::string v)
 : IValue(ivalue::ConstantString::create(std::move(v))) {}
 
-inline IValue::IValue(c10::List<double> v)
-: tag(Tag::DoubleList), is_intrusive_ptr(true) {
-  payload.as_intrusive_ptr = v.impl_.release();
-}
-inline IValue::IValue(std::vector<double> v)
-: IValue(c10::impl::toList(std::move(v))) {}
-
-inline IValue::IValue(c10::List<bool> v)
-: tag(Tag::BoolList), is_intrusive_ptr(true) {
-  payload.as_intrusive_ptr = v.impl_.release();
-}
-inline IValue::IValue(std::vector<bool> v)
-: IValue(c10::impl::toList(std::move(v))) {}
-
-inline IValue::IValue(c10::List<at::Tensor> v)
-: tag(Tag::TensorList), is_intrusive_ptr(true) {
-  payload.as_intrusive_ptr = v.impl_.release();
-}
-inline IValue::IValue(std::vector<at::Tensor> v)
-: IValue(c10::impl::toList(std::move(v))) {}
 
 inline IValue::IValue(c10::impl::GenericList v)
 : tag(Tag::GenericList), is_intrusive_ptr(true) {
   payload.as_intrusive_ptr = v.impl_.release();
 }
 
 template<class T> inline IValue::IValue(c10::List<T> v)
-: IValue(impl::toGenericList<T>(std::move(v))) {
-  static_assert(std::is_same<IValue, typename c10::List<T>::StorageT>::value, "Can only use this constructor for generic list types");
+: IValue(impl::toGenericList<T>(std::move(v))) {}
+template<class T> inline IValue::IValue(at::ArrayRef<T> v)
+: IValue(c10::List<T>()) {
+  auto list = to<c10::List<T>>();
+  list.reserve(v.size());
+  for (const auto& e : v) {
+    list.push_back(std::move(e));
+  }
 }
-template<class T> inline IValue::IValue(std::vector<T> v)
+template<class T> inline IValue::IValue(const std::vector<T>& v)
```

Comment:
This improves performance for people that pass in a read-only vector, because they don't need to copy the vector anymore, but it regresses performance for people who pass in an rvalue vector, because the vector elements can't be moved anymore. You can get the best for both cases by also adding a `std::vector<T>&&` overload.

Reply:
Because T=Tensor (e.g.) but it is internally IValue, this won't prevent a new list being created. It does prevent reference counts. However, there are annoying issues with std::vector<bool> specialization that make it even more complicated to do this. If we end up with this being a perf concern we can implement it.